### PR TITLE
Add ES Modules

### PR DESCRIPTION
**Pull request by tani**
This pull request adds 'latex.esm.js' which exports global object with ES Modules.
We can use following way if this request is marged.

```html
  <script type="module">
  import latexjs from "https://cdn.jsdelivr.net/npm/latex.js/dist/latex.component.js"
  customElements.define('latex-js', latexjs)
  </script>
```

and

```html
  <script type="module">
  import latexjs from "https://cdn.jsdelivr.net/npm/latex.js/dist/latex.esm.js"
  const { parse, HtmlGenerator } = latexjs;
  // blah blah blah...
  </script>
```

### CODE REVIEW

**Comment by fgborges** [comment]
This is off-topic. Since I'm very attracted by this project, I made so so many changes in this week.
Then I decide to fork and start new project because those changes includes changes breaks some compatibility of this project. However, I'll keep to share my contribution to your project!
If you want, off course, take them freely :wink: https://github.com/fgborges/latex2e-js

**Comment by michael-brade** [comment]
Good idea! 

A few things:

* please use `package.json.sh` for changes, `package.json` is a generated file (by executing `package.json.sh`)
* why not call `customElements.define('latex-js', latexjs)` in the `latex.component.js`? Do we need the flexibility but making the user always call `customElements.define()`?
* can you please add a one-line comment of why you made the change in `latex.ltx.ls`? This change would mean we cannot add more document classes without actually changing `latex.ltx.ls`, so I'd only add this if absolutely necessary

**Comment by michael-brade** [comment]
to your off-topic comment: great, that's the idea! I just took a quick look, but I'm not sure what you meant with "breaks compatibility". Any comments? :grinning: I don't think there is any compatibility that has to be kept, as long as the test cases still run. If you meant switching from webpack to rollup--no problem, I'm not attached to webpack, it was just the tool that worked for me. If there is a reason not to use it, I'm all for it.

**Comment by fgborges** [comment]
>     * please use `package.json.sh` for changes, `package.json` is a generated file (by executing `package.json.sh`)

I got it. I update it soon.

>     * why not call `customElements.define('latex-js', latexjs)` in the `latex.component.js`? Do we need the flexibility but making the user always call `customElements.define()`?

Because  the class might be used in the other framework such as React, Vue, and Angular. 
Thus, users does not always calls `customElements.define`. And also, that API defines globally in the page but the concept of ES Modules expects the modules working in the closed environment, e.g., the toplevel variables couldn't be referenced outside of the module without exporting it.

On the other hand, the previous version of `latex.component.js` would runs without ES Modules, that is, we never mind to define globally, I thought. 

>     * can you please add a one-line comment of why you made the change in `latex.ltx.ls`? This change would mean we cannot add more document classes without actually changing `latex.ltx.ls`, so I'd only add this if absolutely necessary

Because rollup cannot treat __dynamic require__ not like webpack.

**Comment by michael-brade** [comment]
> > ```
> > * please use `package.json.sh` for changes, `package.json` is a generated file (by executing `package.json.sh`)
> > ```
> 
> I got it. I update it soon.

Great, thanks :)
 
> Because rollup cannot treat **dynamic require** not like webpack.

Oh... So then why use rollup at all? That's a serious reason not to use it.

**Comment by fgborges** [comment]
You can switch to rollup from webpack? that is good news for me.
One of the biggest reason is __WebWorker support__. I'll add DOM emulation library (svgdom, JSDOM, etc) because we cannot touch DOM in the worker script. That makes the file size 2 times from original. And also, I want to remove `.stylesAndScripts` API and combine assets such as fonts and stylesheet into one js file as dataURI to inject the assets to the output in the generator process.

> Oh... So then why use rollup at all? That's a serious reason not to use it.

Webpack has no support for ES Module.
Instead of that, we may be able to use [dynamic module import API](https://v8.dev/features/dynamic-import)

**Comment by fgborges** [comment]
`package.json.sh` is updated.

**Comment by michael-brade** [comment]
> You can switch to rollup from webpack? that is good news for me.

sure :-) If you get it to work, I'll merge it!

> One of the biggest reason is **WebWorker support**. I'll add DOM emulation library (svgdom, JSDOM, etc) because we cannot touch DOM in the worker script. That makes the file size 2 times from original.

I would recommend `svgdom`. It's small and supports exactly the features LaTeX.js needs. But two questions:
* what do think will the usecase of WebWorkers be? Several LaTeX.js components on one page?
* doesn't webpack support WebWorkers just as rollup does?

> And also, I want to remove `.stylesAndScripts` API and combine assets such as fonts and stylesheet into one js file as dataURI to inject the assets to the output when the generator process.

Interesting idea. It may be that you will have to use Blobs, though, because Chrome supports only 2 MB for data URLs (see https://stackoverflow.com/a/41755526). So maybe you could add a CLI argument, like `--assets-inline` or `--inline` or so, then we can support both use cases.

> > Oh... So then why use rollup at all? That's a serious reason not to use it.
> 
> Webpack has no support for ES Module, but rollup can do it.
> Instead of that, we may be able to use [dynamic module import API](https://v8.dev/features/dynamic-import)

Indeed, I was just reading about that before seeing your comment. Well, as I said, if it works with the current set of features (like it allows peolpe to add custom document classes and packages), go for it!

**Comment by michael-brade** [comment]
One more thing: did you forget to add the `rollup.config.js`? 

And a minor one: after editing `package.json.sh`, please always execute it (`./package.json.sh`) to update the actual `package.json`.

**Comment by fgborges** [comment]
Oops, I add it now.

**Comment by fgborges** [comment]
- Even if  the markdown, some people hoping render it faster with WebWorker. e.g., https://dev.to/mizchi/off-thread-markdown-rendering-with-comlink-2hfo 
- A Japanese said, about LaTeX.js, "It's not so fast, and a bit slow, we hope more faster with WebAssembly. (変換負荷は小さくなく、若干重たいという難点はあります。WebAssemblyになったりするともの凄く快適に使えそうです。)", https://www.moongift.jp/2018/06/latex-js-web%E3%83%96%E3%83%A9%E3%82%A6%E3%82%B6%E4%B8%8A%E3%81%A7%E5%8B%95%E4%BD%9C%E3%81%99%E3%82%8Blatex%E5%9F%B7%E7%AD%86%E7%92%B0%E5%A2%83/

WebWorker makes this project better, and this feature does not relate the issue that is webpack or not.

- I plan to use (rollup-plugin-rul)[https://github.com/rollup/rollup-plugin-url] and i'll add option to CLI.
- I try to keep `documentclass` function of LaTeX.js, give me a day to make a snippet.

**Comment by michael-brade** [comment]
Cool, that's nice to see someone from the other side of the world write about this little project :) But very funny, LaTeX.js is significantly faster than LaTeX itself, taking less than a second to load and only a few tens of milliseconds to render. Still, they call it slow :-D Anyway, I'm always open for optimization. If it is possible to make it faster, let's do it. I just have my doubts about WebWorkers, I cannot imagine to reduce rendering by just using them... I will be very curious about what you are going to create :grinning:

**Comment by fgborges** [comment]
I see. Those topic is out of this request because this request is named "Add ES Modules". I will add the dynamic import capability to this request in a few hours.

**Comment by fgborges** [comment]
I report the changes:

# Before (in master branch)

- Webpack and node.js bundle all files in the documentclass directory into one, and user can be load the files in the directory
- In the browser, users __cannot__ load other documentclasses outside of the documentclass directory without building the library again, because webpack detects files which should be bundled by the form `require "./documentclass/"+documentclass+".js"`

# After (in this request)

- node.js can load any documentclasses in that directory like master branch. i.e., __there is nothing changed before.__
- rollup bundles all files listed in the `latex.ltx.ls` and user can load all files in the documentclasses if they are listed.
- In the browser, users also __cannot__ load other documentclasses outside of the documentclass directory without building the library again,

# Difference

CLI works with same behavior, but in the browser:
- Additional step: users add lines in `latex.ltx.ls` if they want to use the custom documentclass.
- Same step:  in the browser, that is, users should build the library again because of the limitation of webpack and rollup.

# Conclusion
- In node.js I keeps the dinamic custom documentclass feature.
- In the browser, we have had no such kinds of feature even if the master branch. you may know, we have to build the library again.

How do you feel about the additional step? I feel that the no so big change compared to rebuilding.

**Comment by michael-brade** [comment]
Lovely! I agree, you are right. 

Now, GitHub doesn't let me merge this due to conflicts, and clicking Resolve conflicts gets me an eternal spinner... can you please rebase your branch to master?

**Comment by michael-brade** [comment]
I noticed one more thing: `npm run build` does `npm run devbuild` and additionally minifies the result for production. Your new `latex.esm.js` won't be minified by the build yet. Do you know if it is possible to use rollup to minify, or do we need to use terser manually? 

(I also discovered that I am not doing it properly with webpack... will fix this soon)

File: src/latex.ltx.ls [hunk 1]
```diff
@@ -1186,7 +1186,12 @@ export class LaTeX
         @\documentclass = !-> @g.error "Two \\documentclass commands. The document may only declare one class."
 
         # load and instantiate the documentclass
-        Export = (require "./documentclasses/" + documentclass + ".js")
+        Classes = 
+            (documentclass) :-> require ("./documentclasses/"+documetclass+".js")
+            \article        :-> require "./documentclasses/article.js"
+            \book           :-> require "./documentclasses/book.js"
+            \report         :-> require "./documentclasses/article.js"
```
Comment (michael-brade) [inline src/latex.ltx.ls]: copy&paste error: report loads article

File: README.md [hunk 2]
```diff
@@ -94,8 +94,10 @@ LaTeX.js can be used as a web component:
 
   <meta name="viewport" content="width=device-width, initial-scale=1.0">
 
-  <script src="https://cdn.jsdelivr.net/npm/latex.js@0.11.1/dist/latex.min.js"></script>
-  <script src="https://cdn.jsdelivr.net/npm/latex.js@0.11.1/dist/latex.component.js"></script>
+  <script type="module">
+  import latexjs from "https://cdn.jsdelivr.net/npm/latex.js/dist/latex.component.js"
+  customElements.define('latex-js', latexjs)
+  </script>
```
Comment (michael-brade) [inline README.md]: running `npm test` fails for the component, so you also need to update `test/api/webcomponent.html`...

**Review by michael-brade — CHANGES_REQUESTED** [review]

**Comment by fgborges** [comment]
@michael-brade I followed your change request :wink:

**Comment by fgborges** [comment]
> I noticed one more thing: `npm run build` does `npm run devbuild` and additionally minifies the result for production. Your new `latex.esm.js` won't be minified by the build yet. Do you know if it is possible to use rollup to minify, or do we need to use terser manually?

I got it! Wait a minute

**Comment by fgborges** [comment]
Yes, I can add minified version 
However, jsDelivr has server side minifying feature,

> // add ".min" to any JS/CSS file to get a minified version
> // if one doesn't exist, we'll generate it for you
> https://cdn.jsdelivr.net/npm/jquery@3.2.1/src/core.min.js

We may not need to upload minified version.

**Comment by michael-brade** [comment]
> Yes, I can add minified version
> However, jsDelivr has server side minifying feature,
> 
> > // add ".min" to any JS/CSS file to get a minified version
> > // if one doesn't exist, we'll generate it for you
> > https://cdn.jsdelivr.net/npm/jquery@3.2.1/src/core.min.js
> 
> We may not need to upload minified version.

Hm.... interesting. If that were the only way to get the code, great. But I don't know... what about people that use it locally? the less code they have to load, the better and the faster (I think). I always try to make the npm packages as small as possible. It's strange that if you write a small app, you have to drag 50-100 MB of `node_modules` along.

Two more comments (sorry... so many ;-)

* why did you add a server for the tests? I would love to just have the tests run with as little setup as possible. 
* I think you accidentally changed `package.json`, or you did not execute `package.json.sh` after updating it...

**Comment by fgborges** [comment]
> why did you add a server for the tests? I would love to just have the tests run with as little setup as possible.

ES Modules requires the right MIME type header in the `import` syntax. `file://` protocol makes error in the code. Since, I use 'http' protocol in the test.

> I think you accidentally changed package.json, or you did not execute package.json.sh after updating it...

Oops, okay.

**Comment by michael-brade** [comment]
> > why did you add a server for the tests? I would love to just have the tests run with as little setup as possible.
> 
> ES Modules requires the right MIME type header in the `import` syntax. `file://` protocol makes error in the code. Since, I use 'http' protocol in the test.

I see. Strange... it doesn't work for me. No idea why, though.

**Comment by fgborges** [comment]
Please try it again. I forgot to remove some line when I merged them.

**Review by michael-brade — APPROVED** [review]

**Comment by fgborges** [comment]
@michael-brade Finally, I am sure that the es module should not be minified, because vue doesn't provide that kind of versions but they provides just only unminified version. We should follow the large community, don't you? see also https://vuejs.org/v2/guide/installation.html.

**Pull request merged**